### ISSUE TITLE
Groovy plugin script cannot resolve IntelliJ and Live Plugin API classes

### ISSUE THREAD

**Issue by kriegaex** [dkandalov/live-plugin#175]
I just installed the plugin for the first time in IDEA 2023.3.6 Ultimate and opened the Groovy "hello world" example. I can successfully run the plugin, but in the editor it looks like this:

![image](https://github.com/dkandalov/live-plugin/assets/1537384/410f342a-8be4-4b47-bb41-bfbfed8d75d1)

Is there anything else I need to configure?

**Comment by kriegaex** [comment]
It seems as if this helps:

![image](https://github.com/dkandalov/live-plugin/assets/1537384/e4180ef7-def3-4112-927d-ec9627feb4dd)

But unfortunately, I have literally hundreds of projects in my workspace. Furthermore, whenever e.g. Maven is the leading build tool, each automatic or manual re-import resets the dependencies, i.e. I have to start over and re-add the APi and plugin libraries over and over.

Please advise me how to set this up in a better and more persistent way. Thank you. P.S.: I do not speak Kotlin, which is the reason why I chose Groovy.

**Comment by dkandalov** [comment]
> Is there anything else I need to configure?

Not really. Groovy highlighting in the editor has always been a bit of a hack to make a point that it would be nice to have it. It seems like there are APIs in IntelliJ to achieve that but I never got around to figuring it out how to do that for Groovy (and I switched to using Kotlin a few years ago).

You could create a separate project for your live plugin with all the dependencies and it will have syntax highlighting 🤔 Each live plugin is just a folder and you can find the path with the copy path action `ctrl+shift+c` in the tool window or in project view scratches.

Or once the plugin becomes more established I sometimes create a "proper plugin" project with plugin.xml but still keep it reloadable via live plugin for faster feedback by adding plugin jar to live plugin classpath (https://github.com/dkandalov/live-plugin/blob/master/plugin-examples/groovy/additional-classpath/plugin.groovy).

**Comment by kriegaex** [comment]
> adding plugin jar to live plugin classpath (https://github.com/dkandalov/live-plugin/blob/master/plugin-examples/groovy/additional-classpath/plugin.groovy).

Well, there is the next problem. Globbing does not work as advertised. This fails:

```groovy
// add-to-classpath $HOME/.m2/repository/org/mockito/mockito-core/3.3.3/mockito*.jar
```

I only got the plugin working with the full path:

```groovy
// add-to-classpath $HOME/.m2/repository/org/mockito/mockito-core/3.3.3/mockito-core-3.3.3.jar
```

**Comment by kriegaex** [comment]
Another issue is that the imports do not seem to work. I wanted to convert [this Kotlin live plugin](https://youtrack.jetbrains.com/issue/IDEA-193838/Enable-post-compile-weave-mode-for-all-projects-in-one-place#focus=Comments-27-9093453.0-0) to Groovy and extend it, but some plugin classes are not found:

![image](https://github.com/dkandalov/live-plugin/assets/1537384/d61eea3f-c8b4-445e-a81c-a3394a6e0fe9)

**Comment by dkandalov** [comment]
Hm, this works fine for me, maybe it's broken on Windows 🤦 
```
// add-to-classpath $HOME/.m2/repository/org/mockito/mockito-core/5.7.0/mockito*.jar
show(org.mockito.Mockito.class.name)
```

Agreed about butchered line breaks.

Looking at the errors:
 - `unable to resolve class liveplugin.PluginUtil.showInConsole` this is odd because there is such function https://github.com/dkandalov/live-plugin/blob/92932d628f81fe86519663994a7d267d22695f28/src/plugin-api-groovy/liveplugin/PluginUtil.groovy#L150 I wonder if it's because of the way it's used or something else is wrong
 - `unable to resolve class liveplugin.show` should be `import static liveplugin.PluginUtil.show`
 - `unable to resolve class liveplugin.whenDisposed` there is no such function in Groovy API, but the same thing ca be expressed with `liveplugin.implementation.Misc#newDisposable(com.intellij.openapi.Disposable, groovy.lang.Closure)`

I would approach converting Kotlin plugin to Groovy by rewriting it in small incremental steps.

I didn't really think about the case of converting Kotlin live plugin to Groovy 🤔
No problem not using Kotlin 😅 Initially, LivePlugin was Groovy only (because Kotlin didn't exist). But now IntelliJ is being converted to Kotlin anyway so it seems more logical to use it.

Ideally, LivePlugin APIs would be the same or very similar in both Groovy and Kotlin, but the Groovy one was first and I learned few things since then. Besides, LivePlugin API is a small wrapper around IntelliJ APIs which is quite big so it's impossible to cover it all and maintain it. The end goal would be to convince Jetbrains to maintain nice to use reloadable APIs (and some of them are!).

**Comment by kriegaex** [comment]
> Hm, this works fine for me, maybe it's broken on Windows 🤦

Only, if you used platform-specific tools for implementing it, such as shell scripts.

> I would approach converting Kotlin plugin to Groovy by rewriting it in small incremental steps.

Of course, this is exactly what I started to do. But it already failed at the very first step, importing classes from other plugins and resolving them, without even any active plugin code. Not knowing Kotlin, I did not notice that there obviously you do not need `static` for importing methods, so it did not occur to me that after copy & paste I would need to change that in Groovy. Now I know, thanks. I got the plugin and some of my own changes running. 😊

> Ideally, LivePlugin APIs would be the same or very similar in both Groovy and Kotlin, but the Groovy one was first and I learned few things since then.

Does that mean, Groovy support remains as-is and you are hinting me at switching to Kotlin, or are you going to adjust the Groovy part, making it comparable in its capabilities to the Kotlin ones?

**Comment by kriegaex** [comment]
Hm, something strange is happening. After starting the plugin, it does exactly what it is supposed to, but it also blocks part of the IDE, e.g. I can delete, but not enter any text in the editor anymore and also cannot close the IDE without killing the task. Other parts of the IDE remain functional.

Probably I have a silly miskate in the plugin code, but what could it be? Even unloading the plugin does not unblock the IDE.

**Comment by dkandalov** [comment]
> Does that mean, Groovy support remains as-is and you are hinting me at switching to Kotlin, or are you going to adjust the Groovy part, making it comparable in its capabilities to the Kotlin ones?

I wasn't hinting at switching to Kotlin 😅 Although my next plans for live plugin would be about integrating proper IDE runner, configurable compile and runtime classpaths, maybe built-in PSI viewer, object inspector, plugin reload while typing, etc. I haven't considered revisiting Groovy API, it's the first time anyone asked about it 🤔

**Comment by kriegaex** [comment]
Well, I love to write tests in Spock (or Geb, if they are web UI tests), therefore I know some Groovy. Using that language, I can get live plugin stuff done without learning a new language first. Comfort zone...

**Comment by mikehearn** [comment]
This isn't Groovy specific. I installed LivePlugin today and it fails in the same way for both Kotlin and Groovy. For some reason the class path of the plugin isn't correct for the editor. For running it works fine!

**Comment by mikehearn** [comment]
This is IJ Ultimate, latest version, and LivePlugin was installed from the marketplace yesterday. So presumably it's the latest version.

**Comment by pauliesnug** [comment]
@dkandalov I have the same issue as @mikehearn on IntelliJ IDEA Ultimate
